File: MoonEngine-Root/Base/src/main/java/com/greentree/engine/moon/base/assets/scene/XMLSceneAssetSerializator.java
package com.greentree.engine.moon.base.assets.scene;

import com.greentree.commons.data.resource.Resource;
import com.greentree.commons.reflection.ClassUtil;
import com.greentree.commons.reflection.info.TypeInfo;
import com.greentree.commons.util.string.RefStringBuilder;
import com.greentree.commons.xml.XMLElement;
import com.greentree.engine.moon.assets.asset.Asset;
import com.greentree.engine.moon.assets.asset.AssetKt;
import com.greentree.engine.moon.assets.asset.Value1Function;
import com.greentree.engine.moon.assets.key.AssetKey;
import com.greentree.engine.moon.assets.key.ResourceAssetKey;
import com.greentree.engine.moon.assets.key.ResultAssetKey;
import com.greentree.engine.moon.assets.serializator.AssetSerializator;
import com.greentree.engine.moon.assets.serializator.manager.AssetManager;
import com.greentree.engine.moon.base.component.AnnotationUtil;
import com.greentree.engine.moon.base.layer.Layer;
import com.greentree.engine.moon.base.name.Name;
import com.greentree.engine.moon.base.name.Names;
import com.greentree.engine.moon.base.parent.Parent;
import com.greentree.engine.moon.ecs.Entity;
import com.greentree.engine.moon.ecs.World;
import com.greentree.engine.moon.ecs.component.Component;
import com.greentree.engine.moon.ecs.scene.Scene;
import com.greentree.engine.moon.ecs.scene.SceneProperties;
import com.greentree.engine.moon.ecs.scene.SimpleScene;
import com.greentree.engine.moon.ecs.scene.WorldProperty;
import com.greentree.engine.moon.ecs.system.DestroySystem;
import com.greentree.engine.moon.ecs.system.ECSSystem;
import com.greentree.engine.moon.ecs.system.InitSystem;
import com.greentree.engine.moon.ecs.system.UpdateSystem;
import com.greentree.engine.moon.ecs.system.debug.DebugSystems;
import com.greentree.engine.moon.ecs.system.debug.PrintStreamSystemsProfiler;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class XMLSceneAssetSerializator implements AssetSerializator<Scene> {

    private static final Logger LOG = LogManager.getLogger(XMLSceneAssetSerializator.class);

    @Override
    public boolean canLoad(AssetManager manager, AssetKey key) {
        return manager.canLoad(XMLElement.class, key);
    }

    @Override
    public Asset<Scene> load(AssetManager manager, AssetKey ckey) {
        if (manager.canLoad(XMLElement.class, ckey)) {
            final var res = manager.load(XMLElement.class, ckey);
            return AssetKt.map(res, new XMLWorldFunction(manager));
        }
        return null;
    }

    @SuppressWarnings("unchecked")
    private static <T> Class<T> findClass(Class<T> baseClass, String className) throws ClassNotFoundException {
        return (Class<T>) ClassUtil.loadClassInAllPackages(baseClass, className);
    }

    private static final class XMLWorldFunction implements Value1Function<XMLElement, Scene> {

        private static final long serialVersionUID = 1L;
        private final ObjectXMLBuilder builder = new ObjectXMLBuilder();
        private final AssetManager context;
        private final Map<String, Entity> names = new HashMap<>();

        public XMLWorldFunction(AssetManager context) {
            this.context = context;
            builder.add(new XMLTypeAddapter() {

                @Override
                public <T> Constructor<T> newInstance(Context c, TypeInfo<T> type, XMLElement xml_value) {
                    final var xml_value_text = xml_value.getContent();
                    final var key = new ResultAssetKey(xml_value_text);
                    if (context.canLoad(type, key)) {
                        final var v = context.load(type, key).getValue();
                        return new ValueConstructor<>(v);
                    }
                    return null;
                }
            });
            builder.add(new XMLTypeAddapter() {

                @SuppressWarnings("unchecked")
                @Override
                public <T> Constructor<T> newInstance(Context c, TypeInfo<T> type, XMLElement xml_value) {
                    if (ClassUtil.isExtends(Asset.class, type.toClass())) {
                        final var value_type = type.getTypeArguments()[0].getBoxing();
                        try (final var key = c.newInstance(AssetKey.class, xml_value)) {
                            final var value = context.load(value_type, key.value());
                            return new ValueConstructor<>((T) value);
                        }
                    }
                    return null;
                }

                @Override
                public Class<?> getLoadOnly() {
                    return Asset.class;
                }
            });
            builder.add(new XMLTypeAddapter() {

                @Override
                public <T> Constructor<T> newInstance(Context c, TypeInfo<T> type, XMLElement xml_value) {
                    var iter = xml_value.getChildrens();
                    if (iter.isEmpty())
                        return null;
                    final var key = new ResultAssetKey(xml_value.getChildrens().iterator().next());
                    if (context.canLoad(type, key)) {
                        final var v = context.load(type, key).getValue();
                        return new ValueConstructor<>(v);
                    }
                    return null;
                }
            });
            builder.add(new XMLTypeAddapter() {

                @SuppressWarnings("unchecked")
                @Override
                public <T> Constructor<T> newInstance(Context context, TypeInfo<T> type, XMLElement xml_value) {
                    if (AssetKey.class == type.toClass()) {
                        final var xml_value_text = xml_value.getContent();
                        final var xml_value_type = xml_value.getAttribute("type");
                        final AssetKey key;
                        if (xml_value_type == null) {
                            key = new ResourceAssetKey(xml_value_text);
                        } else {
                            final Class<AssetKey> cls;
                            try {
                                cls = findClass(AssetKey.class, xml_value_type);
                            } catch (ClassNotFoundException e) {
                                throw new IllegalArgumentException(e);
                            }
                            try (final var c = context.newInstance(cls, xml_value)) {
                                key = c.value();
                            }
                        }
                        return new ValueConstructor<>((T) key);
                    }
                    return null;
                }

                @Override
                public Class<?> getLoadOnly() {
                    return AssetKey.class;
                }
            });
            builder.add(new XMLTypeAddapter() {

                @Override
                public <T> Constructor<T> newInstance(Context context, TypeInfo<T> type, XMLElement xml_value) {
                    if (!ClassUtil.isExtends(Entity.class, type.toClass()))
                        return null;
                    var name = xml_value.getContent();
                    return () -> (T) names.get(name);
                }

            });
        }

        @Override
        public Scene apply(XMLElement xml_scene) {
            names.clear();
            return new SimpleScene() {

                @Override
                public void build(SceneProperties properties) {
                    var world = properties.get(WorldProperty.class).getWorld();
                    var xml_entities = new ArrayList<XMLElement>();
                    for (var xml_entity_ref : xml_scene.getChildrens("entity_ref"))
                        xml_entities.add(buildEntityRef(xml_entity_ref));
                    for (var xml_entity : xml_scene.getChildrens("entity"))
                        xml_entities.add(xml_entity);
                    var iter = xml_entities.iterator();
                    while (iter.hasNext()) {
                        var xml_entity = iter.next();
                        try {
                            addEntity(properties, world, xml_entity);
                            iter.remove();
                        } catch (RuntimeException e) {
                        }
                    }
                    for (var xml_entity : xml_entities)
                        addEntity(properties, world, xml_entity);
                }

                @Override
                public ECSSystem getSystems(Iterable<? extends ECSSystem> globalSystems) {
                    final var initSystems = new ArrayList<InitSystem>();
                    final var updateSystems = new ArrayList<UpdateSystem>();
                    final var destroySystems = new ArrayList<DestroySystem>();
                    for (var system : globalSystems) {
                        if (system instanceof InitSystem s)
                            initSystems.add(s);
                        if (system instanceof UpdateSystem s)
                            updateSystems.add(s);
                        if (system instanceof DestroySystem s)
                            destroySystems.add(s);
                    }
                    for (var xml_system : xml_scene.getChildrens("system"))
                        try {
                            var system = newFromXML(ECSSystem.class, xml_system);
                            if (system instanceof InitSystem s)
                                initSystems.add(s);
                            if (system instanceof UpdateSystem s)
                                updateSystems.add(s);
                            if (system instanceof DestroySystem s)
                                destroySystems.add(s);
                        } catch (RuntimeException | ClassNotFoundException e) {
                            LOG.error("", e);
                        }
                    final var log = new File("log");
                    log.mkdirs();
                    AnnotationUtil.sortInit(initSystems);
                    AnnotationUtil.sortUpdate(updateSystems);
                    AnnotationUtil.sortDestroy(destroySystems);
                    return new DebugSystems(
                            new PrintStreamSystemsProfiler(new File(log, "systems_init.log"),
                                    new File(log, "systems_update.log"), new File(log, "systems_destroy.log")),
                            initSystems, updateSystems, destroySystems);
                }

                private <T> T newFromXML(Class<T> baseClass, XMLElement xml_element) throws ClassNotFoundException {
                    final var systemClassName = xml_element.getAttribute("type");
                    final var cls = findClass(baseClass, systemClassName);
                    try (final var c = builder.newInstance(cls, xml_element)) {
                        if (c == null)
                            return null;
                        return c.value();
                    }
                }

                private XMLElement buildEntityRef(XMLElement xml_entity_ref) {
                    final var file = xml_entity_ref.getAttribute("file");
                    final var res = context.load(Resource.class, file).getValue();
                    final RefStringBuilder builder;
                    try {
                        builder = RefStringBuilder.build(res.open());
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                    final var map = new HashMap<String, String>();
                    for (var xmLproperty : xml_entity_ref.getChildrens("property")) {
                        final var name = xmLproperty.getAttribute("name");
                        final var value = xmLproperty.getContent();
                        map.put(name, value);
                    }
                    final var text = builder.toString(map);
                    return context.load(XMLElement.class, new ResultAssetKey(text)).getValue();
                }

                private Entity addEntity(SceneProperties properties, World world, XMLElement xml_entity) {
                    final var name_atr = xml_entity.getAttribute("name");
                    final var layer_atr = xml_entity.getAttribute("layer");
                    final var parent_atr = xml_entity.getAttribute("parent");
                    final var entity = world.newEntity();
                    try (final var lock = entity.lock()) {
                        if (name_atr != null) {
                            names.put(name_atr, entity);
                            lock.add(new Name(name_atr));
                        }
                        if (layer_atr != null)
                            lock.add(new Layer(layer_atr));
                        for (var xml_component : xml_entity.getChildrens("component"))
                            try {
                                final var c = newFromXML(Component.class, xml_component);
                                if (c != null)
                                    lock.add(c);
                            } catch (RuntimeException | ClassNotFoundException e) {
                                LOG.error("", e);
                            }
                    }
                    if (parent_atr != null) {
                        var parent = properties.get(Names.class).get(parent_atr);
                        Parent.setParent(entity, parent);
                    }
                    for (var xml_child : xml_entity.getChildrens("entity")) {
                        Parent.setParent(addEntity(properties, world, xml_child), entity);
                    }
                    for (var xml_child : xml_entity.getChildrens("entity_ref")) {
                        Parent.setParent(addEntity(properties, world, buildEntityRef(xml_child)), entity);
                    }
                    return entity;
                }

            };
        }

    }

}

File: MoonEngine/Base/src/main/java/com/greentree/engine/moon/base/assets/scene/XMLSceneAssetSerializator.kt
package com.greentree.engine.moon.base.assets.scene

import com.greentree.commons.reflection.ClassUtil
import com.greentree.commons.reflection.info.TypeInfo
import com.greentree.commons.xml.XMLElement
import com.greentree.engine.moon.assets.Value1Function
import com.greentree.engine.moon.assets.asset.Asset
import com.greentree.engine.moon.assets.key.AssetKey
import com.greentree.engine.moon.assets.key.ResourceAssetKey
import com.greentree.engine.moon.assets.serializator.AssetSerializator
import com.greentree.engine.moon.assets.serializator.loader.AssetLoader
import com.greentree.engine.moon.assets.serializator.loader.load
import com.greentree.engine.moon.assets.serializator.manager.loadAsync
import com.greentree.engine.moon.base.AssetManagerProperty
import com.greentree.engine.moon.base.assets
import com.greentree.engine.moon.base.assets.scene.adapters.Constructor
import com.greentree.engine.moon.base.assets.scene.adapters.Context
import com.greentree.engine.moon.base.assets.scene.adapters.ObjectXMLBuilder
import com.greentree.engine.moon.base.assets.scene.adapters.ValueConstructor
import com.greentree.engine.moon.base.assets.scene.adapters.XMLTypeAddapter
import com.greentree.engine.moon.base.assets.scene.adapters.findClass
import com.greentree.engine.moon.base.assets.text.RefStringBuilderAssetKey
import com.greentree.engine.moon.base.component.AnnotationUtil
import com.greentree.engine.moon.base.layer.Layer
import com.greentree.engine.moon.base.name.Name
import com.greentree.engine.moon.base.name.Names
import com.greentree.engine.moon.base.parent.Parent
import com.greentree.engine.moon.ecs.Entity
import com.greentree.engine.moon.ecs.World
import com.greentree.engine.moon.ecs.component.Component
import com.greentree.engine.moon.ecs.scene.Scene
import com.greentree.engine.moon.ecs.scene.SceneProperties
import com.greentree.engine.moon.ecs.scene.SimpleScene
import com.greentree.engine.moon.ecs.scene.WorldProperty
import com.greentree.engine.moon.ecs.system.DestroySystem
import com.greentree.engine.moon.ecs.system.ECSSystem
import com.greentree.engine.moon.ecs.system.InitSystem
import com.greentree.engine.moon.ecs.system.UpdateSystem
import com.greentree.engine.moon.ecs.system.debug.DebugSystems
import com.greentree.engine.moon.ecs.system.debug.PrintStreamSystemsProfiler
import org.apache.logging.log4j.LogManager
import java.io.File

object XMLSceneAssetSerializator : AssetSerializator<Scene> {

	override fun load(manager: AssetLoader.Context, key: AssetKey): Asset<Scene> {
		val res = manager.load<XMLElement>(key)
		return res.map(XMLWorldFunction)
	}

	private object XMLWorldFunction : Value1Function<XMLElement, Scene> {

		override fun apply(xml_scene: XMLElement): Scene {
			return object : SimpleScene {
				val builder = ObjectXMLBuilder()
				val names: MutableMap<String, Entity> = HashMap()
				override fun build(properties: SceneProperties) {
					val context = properties.get(AssetManagerProperty::class.java).manager
					builder.add(object : XMLTypeAddapter {
						override fun <T> newInstance(
							c: Context,
							type: TypeInfo<T>,
							xml_value: XMLElement,
						): Constructor<T>? {
							if(ClassUtil.isExtends(Asset::class.java, type.toClass())) {
								if(type.typeArguments.size == 0) throw UnsupportedOperationException("asset type without Type Arguments")
								val value_type = type.typeArguments[0].boxing
								c.newInstance(AssetKey::class.java, xml_value).use { key ->
									val value = context.loadAsync(value_type, key.value())
									if(!value.isValid()) {
										value.value
										throw UnsupportedOperationException("build not valid asset $value from $xml_value")
									}
									return ValueConstructor(value as T)
								}
							}
							return null
						}

						override fun getLoadOnly(): Class<*> {
							return Asset::class.java
						}
					})
					builder.add(object : XMLTypeAddapter {
						override fun <T> newInstance(
							context: Context,
							type: TypeInfo<T>,
							xml_value: XMLElement,
						): Constructor<T>? {
							if(!ClassUtil.isExtends(Entity::class.java, type.toClass())) return null
							val name = xml_value.content
							return Constructor { names[name] as T? }
						}
					})
					val world = properties.get(WorldProperty::class.java).world
					val xml_entities = ArrayList<XMLElement>()
					for(xml_entity_ref in xml_scene.getChildrens("entity_ref")) xml_entities.add(
						buildEntityRef(
							properties,
							xml_entity_ref
						)
					)
					for(xml_entity in xml_scene.getChildrens("entity")) xml_entities.add(xml_entity)
					val iter = xml_entities.iterator()
					while(iter.hasNext()) {
						val xml_entity = iter.next()
						try {
							addEntity(properties, world, xml_entity)
							iter.remove()
						} catch(e: RuntimeException) {
						}
					}
					for(xml_entity in xml_entities) addEntity(properties, world, xml_entity)
				}

				override fun getSystems(): ECSSystem {
					val initSystems = ArrayList<InitSystem>()
					val updateSystems = ArrayList<UpdateSystem>()
					val destroySystems = ArrayList<DestroySystem>()
					for(xml_system in xml_scene.getChildrens("system")) try {
						val system = newFromXML(ECSSystem::class.java, xml_system)
						if(system is InitSystem) initSystems.add(system)
						if(system is UpdateSystem) updateSystems.add(system)
						if(system is DestroySystem) destroySystems.add(system)
					} catch(e: RuntimeException) {
						LOG.error("load system", e)
					} catch(e: ClassNotFoundException) {
						LOG.error("load system", e)
					}
					val log = File("log")
					log.mkdirs()
					AnnotationUtil.sortInit(initSystems)
					AnnotationUtil.sortUpdate(updateSystems)
					AnnotationUtil.sortDestroy(destroySystems)
					return DebugSystems(
						PrintStreamSystemsProfiler(
							File(log, "systems_init.log"),
							File(log, "systems_update.log"), File(log, "systems_destroy.log")
						),
						initSystems, updateSystems, destroySystems
					)
				}

				@Throws(ClassNotFoundException::class)
				private fun <T> newFromXML(baseClass: Class<T>, xml_element: XMLElement): T {
					val systemClassName = xml_element.getAttribute("type")
					val cls = findClass(baseClass, systemClassName)
					builder.newInstance(cls, xml_element).use { c ->
						return c.value()
					}
				}

				private fun buildEntityRef(properties: SceneProperties, xml_entity_ref: XMLElement): XMLElement {
					val context = properties.assets
					val file = xml_entity_ref.getAttribute("file")
					val map = HashMap<String, String>()
					for(xmLproperty in xml_entity_ref.getChildrens("property")) {
						val name = xmLproperty.getAttribute("name")
						val value = xmLproperty.content
						map[name] = value
					}
					return context.load(
						XMLElement::class.java,
						RefStringBuilderAssetKey(ResourceAssetKey(file), map)
					).value
				}

				private fun addEntity(properties: SceneProperties, world: World, xml_entity: XMLElement): Entity {
					val name_atr = xml_entity.getAttribute("name")
					val layer_atr = xml_entity.getAttribute("layer")
					val parent_atr = xml_entity.getAttribute("parent")
					val entity = world.newEntity()
					entity.lock().use { lock ->
						if(name_atr != null) {
							names[name_atr] = entity
							lock.add(Name(name_atr))
						}
						if(layer_atr != null) lock.add(Layer(layer_atr))
						for(xml_component in xml_entity.getChildrens("component")) try {
							val c = newFromXML(
								Component::class.java, xml_component
							)
							lock.add(c)
						} catch(e: RuntimeException) {
							LOG.error("load entity", e)
						} catch(e: ClassNotFoundException) {
							LOG.error("load entity", e)
						}
					}
					if(parent_atr != null) {
						val parent = properties.get(Names::class.java)[parent_atr]
						Parent.setParent(entity, parent)
					}
					for(xml_child in xml_entity.getChildrens("entity")) {
						Parent.setParent(addEntity(properties, world, xml_child), entity)
					}
					for(xml_child in xml_entity.getChildrens("entity_ref")) {
						Parent.setParent(addEntity(properties, world, buildEntityRef(properties, xml_child)), entity)
					}
					return entity
				}
			}
		}
	}

	private val LOG = LogManager.getLogger(
		XMLSceneAssetSerializator::class.java
	)
}
